presto-main/src/main/java/com/facebook/presto/sql/Optimizer.java

```diff
@@ -98,6 +101,9 @@ public Plan validateAndOptimizePlan(PlanNode root, PlanStage stage)
         boolean enableVerboseRuntimeStats = SystemSessionProperties.isVerboseRuntimeStatsEnabled(session);
         if (stage.ordinal() >= OPTIMIZED.ordinal()) {
             for (PlanOptimizer optimizer : planOptimizers) {
+                if (Thread.currentThread().isInterrupted()) {
+                    throw new PrestoException(QUERY_PLANNING_TIMEOUT, String.format("The query optimizer exceeded the timeout of %s.", getQueryAnalyzerTimeout(session).toString()));
```

Comment:
Any insight into where the interrupt flag is being set ? Can we throw this exception when the interrupt flag is being set say when the analyzer timeout expires ?

Reply:
Given that the thread can be interrupted by other codepaths as well, is it possible to only fire this when TimeoutThread is invoked ? 

Checking for interrupts is usually not something I have seen done pervasively in the codebase and I am wondering if there is a cleaner way here.

Reply:
> Given that the thread can be interrupted by other codepaths as well, is it possible to only fire this when TimeoutThread is invoked ?
> 
> Checking for interrupts is usually not something I have seen done pervasively in the codebase and I am wondering if there is a cleaner way here.

I believe we do it in a couple of other cases - runaway splits and runaway regex. But we do that slightly differently examining the stack to look for joni classes. Maybe we can do something similar here for consistency though I feel that's hacky as well.

Reply:
I am pretty sure (by doing code search and check each of the `interrupt` call) there are only two place in our own code that we **initiate** the "interrupt" status flag.  One is in the `TimeoutThread` and another one is in `TaskExecutor`. 

We have two case for `TimeoutThread`,  and one is `analyzer.analyzeSemantic`, and another one is `createLogicalPlan`. These two use are using the same timeout setting: query_analyzer_timeout.

All the other places are just responding to an interrupt flag that is already set by someone else.

Reply:
One problem is the code that sets the `interrupt` flag is far away from the code responds to the `interrupt` flag is very far away. It makes debugging harder. Also wondering if there is a better way to do this.

Reply:
> Any insight into where the interrupt flag is being set ? 

The interrupt is set here 
https://github.com/prestodb/presto/blob/614dd6738b05aa0beea06a799b227d2d1c1bf124/presto-main/src/main/java/com/facebook/presto/execution/SqlQueryExecution.java#L448-L454
which will call Thread::interrupt 
https://github.com/prestodb/presto/blob/031ce06fcc3ed898d6bac16a5f2fabaa590f027d/presto-main/src/main/java/com/facebook/presto/execution/TimeoutThread.java#L44
if the enclosed statement in the try block is not complete within timeout.

> Can we throw this exception when the interrupt flag is being set say when the analyzer timeout expires ?

I think this is a good point. I tried to look for ways to achieve this but cannot find a good approach to do it. I am new to Java, and happy to try if there are suggestions on how to do it.

Reply:
This is a correct pattern if we use Thread as primitive, 

I think the better option is to use `Future`/`Task` abstraction, but that requires bigger changes.